### ISSUE TITLE
Refactor: Partial trace and partial transpose

### ISSUE THREAD

**Issue by vprusso** [vprusso/toqito#145]
## The problem:
The present implementation of `partial_trace` and `partial_transpose` is not particularly favorable for a few different reasons:

- Counterintuitively, the first index is "1" and opposed to "0". The violates the conventions used elsewhere and is a result of these functions originally being converted over from MATLAB

- The functions themselves are quite convoluted, and hence, more likely to induce error. While the test suite for these functions is fairly thorough, it is still not ideal. 

## The solution:
The solution is fairly straightforward:

- Replace the `partial_transpose` and `partial_trace` functions with the corresponding equivalent functions provided from the [picos module](https://picos-api.gitlab.io/picos/index.html) (that is, `picos.partial_transpose` and `picos.partial_trace`, respectively. 

- The other places in the code that make use of either `partial_trace` or `partial_tranpose` will necessarily need to be updated along with any mention of usage in the docs/tutorials.

**Comment by ErikQQY** [comment]
Hi there! I am interested in this issue, it would be great if you can assign this issue to me!😄

**Comment by vprusso** [comment]
Thank you for your interest @ErikQQY ! You have been assigned!

**Comment by ErikQQY** [comment]
@vprusso I just refactored ```partial_trace``` and ```partial_transpose```, but as for those input matrix is ```Variable``` matrix generated from cvxpy, ```partial_trace``` and ```partial_transpose``` in picos can't handle them😢, how do we handle these usage?

**Comment by ErikQQY** [comment]
For example:
https://github.com/vprusso/toqito/blob/835fcd3e2b0c44f10ba7877710c80edbd49e3860/toqito/state_opt/optimal_clone.py#L176

https://github.com/vprusso/toqito/blob/835fcd3e2b0c44f10ba7877710c80edbd49e3860/toqito/nonlocal_games/quantum_hedging.py#L131

https://github.com/vprusso/toqito/blob/835fcd3e2b0c44f10ba7877710c80edbd49e3860/toqito/state_opt/ppt_distinguishability.py#L180

**Comment by vprusso** [comment]
> @vprusso I just refactored `partial_trace` and `partial_transpose`, but as for those input matrix is `Variable` matrix generated from cvxpy, `partial_trace` and `partial_transpose` in picos can't handle them😢, how do we handle these usage?

Hi @ErikQQY . Thanks for the work thus far and for the question. Two options come to mind:

1. The more involved solution would be to rewrite the `cvxpy` SDPs to equivalent ones in `picos`. However, I think that's beyond the scope of this issue and is probably overkill for what we need here. 
2. A more prudent approach may be to leverage something like this:

```python
    # If the input matrix is a CVX variable for an SDP, we convert it to a numpy array,
    # perform the partial trace, and convert it back to a CVX variable.
    if isinstance(input_mat, Variable):
        rho_np = expr_as_np_array(input_mat)
        traced_rho = partial_trace(rho_np, sys, dim)
        traced_rho = np_array_as_expr(traced_rho)
        return traced_rho
```

Note that the above snippet of code is used in both the `partial_trace` and `partial_transpose` functions that presently reside in `toqito`. The general idea is that if the type of the input variable is a `cvxpy.Variable` then this helper does some gymnastics to convert it into something that the `partial_*` function can operate on and then converts it back into a cvxpy.Variable` expression. 

I believe the same general approach could also be applied to the portions of the code that you pointed to.

Does that make sense?

**Comment by ErikQQY** [comment]
Yeah, I think the prudent approach is appliable in this case, I think we have two choices:

1. Directly use the above snippet of code in where we use ```partial_trace``` and ```partial_transpose``` with ```cvxpy.Variable```. (But I think this would be a little messy)
2. Instead of directly delete our ```partial_trace``` and ```partial_transpose```, we can still have our own ```partial_trace``` and ```partial_transpose```, but we made a wrapper for ```picos.partial_trace``` and ```picos.partial_transpose```:
```python
def partial_trace(mat, sys, dim):
    if isinstance(input_mat, Variable):
        rho_np = expr_as_np_array(input_mat)
        traced_rho = partial_trace(rho_np, sys, dim)
        traced_rho = np_array_as_expr(traced_rho)
        return traced_rho
    picos.partial_trace(mat, sys, dim)
```

**Comment by vprusso** [comment]
Yep, I think the approach you outlined in (2) sounds like the way to go!

**Comment by ErikQQY** [comment]
@vprusso I complete the ```partial_trace``` refactor, but as for ```partial_transpose```, the usage of  ```partial_transpose``` in toqito is quite different from picos, for example, ```partial_transpose``` in toqito is capable of handling non square matrices, while picos can't:

```python
rho = np.kron(np.eye(2, 3), np.ones((2, 3)))
rho = np.kron(rho, np.eye(2, 3))
dim = np.array([[2, 2, 2], [3, 3, 3]])
res = partial_transpose(rho, sys=2, dim=dim)
```

It confuses me a lot🤔

**Comment by ErikQQY** [comment]
Yeah, I think the partial_transpose should be applied to a square matrix, but as a matter of fact, there is indeed a test for non-square matrices in toqito:

https://github.com/vprusso/toqito/blob/1da8ad2e76c7f7165d2b0d5a8b1d412787a1008f/tests/test_channels/test_partial_transpose.py#L711-L722

**Comment by vprusso** [comment]
Yes, I did indeed see the test for the non-square matrices, however, assuming that the partial transpose is only applicable to square matrices (modulo these tests) I think we can remove these test cases. 

I did not happen to see any other places where applying a partial transpose to a non-square matrix is needed so I think we can:

- Replace the `partial_transpose` function with the `picos.partial_transpose` function
- Deal with the `cvx.Variable` case the same way that you did with the partial trace
- Remove these two unit tests on non-square inputs for the partial transpose
- Ensure all other test cases of the partial transpose pass and that any usages of the partial transpose are properly updated. 
- The docs (for both partial transpose and partial trace) should be updated to reflect this change. 

The above is probably not an exhaustive list, but I think it covers the bulk of it. Does that make sense?

**Comment by ErikQQY** [comment]
Yeah, this make sense to me!

I got another question here:

https://github.com/vprusso/toqito/blob/150aa3ac01f77b3aadc67dd2e32eac6b502168c7/toqito/channels/realignment.py#L80-L84

An MWE can be reproduce using:

```python
a=np.arange(1, 17).reshape(4,4)
partial_transpose(a, 1,  [[2, 2], [2, 2]])
```

As far as I know, picos doesn't have this feature(passing dim as ```[[dim, dim], [dim, dim]]```), how we handle this by replacing with picos? Do we do the same as we did for those cvxpy variables?
